src/main/groovy/com/google/protobuf/gradle/Utils.groovy

```diff
@@ -129,4 +133,78 @@ class Utils {
       }
     }
   }
+
+  /**
+   * Add the the folder of generated source to Eclipse .classpath file.
+   */
+  static void addToEclipseSources(Project project, boolean isTest, String sourceSetName, File f) {
+    project.plugins.withId("eclipse") {
+      File projectDir = project.getProjectDir()
+
+      EclipseModel model = project.getExtensions().findByType(EclipseModel)
+      model.classpath.file.whenMerged { Classpath cp ->
+        // buildship requires the folder exists on disk, otherwise
+        // it will be ignored when updating classpath file, see:
+        // https://github.com/eclipse/buildship/issues/1196
+        f.mkdirs()
+
+        String relativePath = projectDir.toURI().relativize(f.toURI()).getPath()
+        // remove trailing slash
+        if (relativePath.endsWith("/")) {
+          relativePath = relativePath[0..(relativePath.length() - 2)]
+        }
+        SourceFolder entry = new SourceFolder(relativePath, getOutputPath(cp, sourceSetName))
+        entry.entryAttributes.put("optional", "true")
+        entry.entryAttributes.put("ignore_optional_problems", "true")
+
+        // this attribute is optional, but it could be useful for IDEs to recognize that it is
```

Comment:
I'm wary of adding additional information like this. Ideally we wouldn't use an Eclipse API, and we can't have this sort of extra information in a generic API. So having it here leads to one of two possible futures: 1) it is unused and doesn't matter or 2) it is used and now the plugin must always use the Eclipse API.